pkg/updater/download.go

```diff
@@ -0,0 +1,127 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package updater
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+)
+
+const (
+	agentArchiveFileName = "agent.tar.gz"
+)
+
+// downloader is the downloader used by the updater to download packages.
+type downloader struct {
+	client *http.Client
+}
+
+// newDownloader returns a new Downloader.
+func newDownloader(client *http.Client) *downloader {
+	return &downloader{
+		client: client,
+	}
+}
+
+// Download downloads the package at the given URL in temporary directory,
+// verifies its SHA256 hash and extracts it to the given destination path.
+// It currently assumes the package is a tar.gz archive.
+func (d *downloader) Download(ctx context.Context, pkg Package, destinationPath string) error {
+	tmpDir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return fmt.Errorf("could not create temporary directory: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, pkg.URL, nil)
+	if err != nil {
+		return fmt.Errorf("could not create download request: %w", err)
+	}
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("could not download package: %w", err)
+	}
+	defer resp.Body.Close()
+	hashWriter := sha256.New()
+	reader := io.TeeReader(resp.Body, hashWriter)
+	archivePath := filepath.Join(tmpDir, agentArchiveFileName)
+	archiveFile, err := os.Create(archivePath)
+	if err != nil {
+		return fmt.Errorf("could not create archive file: %w", err)
+	}
+	defer archiveFile.Close()
+	_, err = io.Copy(archiveFile, reader)
+	if err != nil {
+		return fmt.Errorf("could not write archive file: %w", err)
+	}
+	sha256 := hashWriter.Sum(nil)
+	expectedHash, err := hex.DecodeString(pkg.SHA256)
+	if err != nil {
+		return fmt.Errorf("could not decode hash: %w", err)
+	}
+	if !bytes.Equal(expectedHash, sha256) {
+		return fmt.Errorf("invalid hash for %s: expected %x, got %x", pkg.URL, pkg.SHA256, sha256)
+	}
+	err = extractTarGz(archivePath, destinationPath)
+	if err != nil {
+		return fmt.Errorf("could not extract archive: %w", err)
+	}
+	return nil
+}
+
+func extractTarGz(archivePath string, destinationPath string) error {
+	f, err := os.Open(archivePath)
+	if err != nil {
+		return fmt.Errorf("could not open archive: %w", err)
+	}
+	defer f.Close()
+	gzr, err := gzip.NewReader(f)
+	if err != nil {
+		return fmt.Errorf("could not create gzip reader: %w", err)
+	}
+	defer gzr.Close()
+	tr := tar.NewReader(gzr)
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("could not read tar header: %w", err)
+		}
+		target := filepath.Join(destinationPath, header.Name)
```

Comment:
This is purposeful, the package itself is trusted in this context as the hash check as been before we start extracting.

With `tar` specifically this is also very tricky because sanitising `header.Name` is not enough since symlinks are supported and can point outside of the directory (I wouldn't be surprised if the agent has a few).

Also keep in mind that we're going to execute that package anyway since it passed the hash check and if I can control its contents as an attacker I have many other ways to get code exec.

If that sounds logical to you I can add this reasoning as a comment in case other people stumble on this and wonder if this is safe.

Reply:
A comment would be great. I think this block may be a good thing to run by SDG team for eval though.

Reply:
>With tar specifically this is also very tricky because sanitising header.Name is not enough since symlinks are supported and can point outside of the directory (I wouldn't be surprised if the agent has a few).

Given that we don't support `TypeSymlink` in the `switch` below I'd assume that we don't support them with the  code as-is

Reply:
SDG does not exist anymore unfortunately 😅. Part of the release process of this feature will include a pentest, I will point them to this part as you are right this file is currently the most sensitive one (hash check especially)

This code is bound to be reworked as the package format is still in design  (see @bkabrda 's comment) so I think I won't immediately add symlink support until that's validated.

Reply:
Added a comment